lib/fizzy/execute.cpp

```diff
@@ -362,64 +406,159 @@ execution_result execute(Instance& instance, FuncIdx function, std::vector<uint6
+        {
+            if (!load_from_memory<uint64_t, int8_t>(instance.memory, stack, immediates))
+            {
+                trap = true;
+                goto end;
+            }
+            break;
+        }
+        case Instr::i64_load8_u:
+        {
+            if (!load_from_memory<uint64_t, uint8_t>(instance.memory, stack, immediates))
+            {
+                trap = true;
+                goto end;
+            }
+            break;
+        }
+        case Instr::i64_load16_s:
+        {
+            if (!load_from_memory<uint64_t, int16_t>(instance.memory, stack, immediates))
+            {
+                trap = true;
+                goto end;
+            }
+            break;
+        }
+        case Instr::i64_load16_u:
+        {
+            if (!load_from_memory<uint64_t, uint16_t>(instance.memory, stack, immediates))
+            {
+                trap = true;
+                goto end;
+            }
+            break;
+        }
+        case Instr::i64_load32_s:
+        {
+            if (!load_from_memory<uint64_t, int32_t>(instance.memory, stack, immediates))
+            {
+                trap = true;
+                goto end;
+            }
+            break;
+        }
+        case Instr::i64_load32_u:
+        {
+            if (!load_from_memory<uint64_t, uint32_t>(instance.memory, stack, immediates))
             {
                 trap = true;
                 goto end;
             }
-            const auto ret = load<uint64_t>(instance.memory, address + offset);
-            stack.push(ret);
             break;
         }
-        // FIXME: make this into a template?
         case Instr::i32_store:
         {
-            const auto value = static_cast<uint32_t>(stack.pop());
-            const auto address = static_cast<uint32_t>(stack.pop());
-            // NOTE: alignment is dropped by the parser
-            const auto offset = read<uint32_t>(immediates);
-            if ((address + offset + sizeof(uint32_t)) > instance.memory.size())
+            if (!store_into_memory<uint32_t>(instance.memory, stack, immediates))
             {
                 trap = true;
                 goto end;
             }
-            store<uint32_t>(instance.memory, address + offset, value);
             break;
         }
-        // FIXME: make this into a template?
         case Instr::i64_store:
         {
-            const auto value = static_cast<uint64_t>(stack.pop());
-            const auto address = static_cast<uint32_t>(stack.pop());
-            // NOTE: alignment is dropped by the parser
-            const auto offset = read<uint32_t>(immediates);
-            if ((address + offset + sizeof(uint64_t)) > instance.memory.size())
+            if (!store_into_memory<uint64_t>(instance.memory, stack, immediates))
+            {
+                trap = true;
+                goto end;
+            }
+            break;
+        }
+        case Instr::i32_store8:
+        case Instr::i64_store8:
```

Comment:
Note: If both instructions remain the same, parser can emit just one of them.

Reply:
I assume that I don't have to change anything. yes?

Reply:
No, you don't have to.

Reply:
I'd suggest we only merge instructions when everything is implemented and we pass the test suite.